src/orange/orangeinp/detail/BuildLogicUtils.hh

```diff
@@ -0,0 +1,342 @@
+#pragma once
+
+#include <algorithm>
+#include <type_traits>
+#include <vector>
+
+#include "corecel/Assert.hh"
+#include "corecel/cont/VariantUtils.hh"
+#include "corecel/math/Algorithms.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/orangeinp/CsgTree.hh"
+#include "orange/orangeinp/CsgTypes.hh"
+
+namespace celeritas
+{
+namespace orangeinp
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Result of building a logic representation of a node.
+ */
+struct BuildLogicResult
+{
+    using VecLogic = std::vector<logic_int>;
+    using VecSurface = std::vector<LocalSurfaceId>;
+
+    VecSurface faces;
+    VecLogic logic;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct a logic representation of a node.
+ *
+ * The result is a pair of vectors: the sorted surface IDs comprising the faces
+ * of this volume, and the logical representation using \em face IDs, i.e. with
+ * the surfaces remapped to the index of the surface in the face vector.
+ *
+ * The function is templated on a policy class that determines the logic
+ * representation. The policy class must have an operator() that takes a
+ * NodeId.
+ *
+ * The per-node local surfaces (faces) are sorted in ascending order of ID, not
+ * of access, since they're always evaluated sequentially rather than as part
+ * of the logic evaluation itself.
+ */
+template<class BuildLogicPolicy>
+inline BuildLogicResult build_logic(BuildLogicPolicy&& policy, NodeId n)
+{
+    static_assert(std::is_invocable_v<BuildLogicPolicy, NodeId>);
+    static_assert(std::is_rvalue_reference_v<BuildLogicPolicy&&>,
+                  "Will move from policy: rvalue ref expected");
+
+    // Construct logic vector as local surface IDs
+    auto& lgc = policy.logic();
+    CELER_EXPECT(lgc.empty());
+    policy(n);
+
+    // Construct sorted vector of faces
+    BuildLogicResult::VecSurface faces;
+    for (auto const& v : lgc)
+    {
+        if (!logic::is_operator_token(v))
+        {
+            faces.push_back(LocalSurfaceId{v});
+        }
+    }
+
+    // Sort and uniquify the vector
+    std::sort(faces.begin(), faces.end());
+    faces.erase(std::unique(faces.begin(), faces.end()), faces.end());
+
+    // Remap logic
+    for (auto& v : lgc)
+    {
+        if (!logic::is_operator_token(v))
+        {
+            auto iter
+                = find_sorted(faces.begin(), faces.end(), LocalSurfaceId{v});
+            CELER_ASSUME(iter != faces.end());
+            v = iter - faces.begin();
+        }
+    }
+    return {faces, std::move(lgc)};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Base class for logic builder policies following CRTP pattern.
+ *
+ * The call operator for Negated and Joined are not implemented in the base
+ * policy and must be provided by the derived class.
+ */
+template<class BuilderPolicy>
```

Comment:
Man I really hate how many suggestions I've made that don't seem write when implemented... I don't think the complexity (and potentially code size, due to having to duplicate the `std::visit` dispatch) of CRTP is justified here. I see one big function that doesn't require a template (`operator()(Surface const&)`), and a couple of trivial functions that could easily be copied.

Maybe it's good as is, but perhaps a small standalone helper class that manages the `mapping_` would be a nice final touch?

Reply:
As you suggested previously, we could replace the `BaseBuildLogicPolicy` inheritance by using composition. It would be simpler but result in more boilerplate as we have to wrap and forward the multiple overloads. 

Regarding code size, I don't think the different options are any different, in any case, we need an implementation of `std::variant` dispatch for `PostfixBuildLogicPolicy` and one for `InfixBuildLogicPolicy`. The alternative would be virtual calls and dynamic dispatch but that has a runtime cost since I assume this is called many times when building the geometry.